src/ImageSharp/Formats/Webp/Lossy/YuvConversion.cs

```diff
@@ -18,6 +23,269 @@ internal static class YuvConversion
+                    YuvToBgr(bottomY[xMul2 - 1], (int)(uv0 & 0xff), (int)(uv0 >> 16), bottomDst.Slice((xMul2 - 1) * xStep));
+                    YuvToBgr(bottomY[xMul2 + 0], (int)(uv1 & 0xff), (int)(uv1 >> 16), bottomDst.Slice((xMul2 + 0) * xStep));
+                }
+
+                tluv = tuv;
+                luv = uv;
+            }
+
+            if ((len & 1) == 0)
+            {
+                uv0 = ((3 * tluv) + luv + 0x00020002u) >> 2;
+                YuvToBgr(topY[len - 1], (int)(uv0 & 0xff), (int)(uv0 >> 16), topDst.Slice((len - 1) * xStep));
+                if (bottomY != default)
+                {
+                    uv0 = ((3 * luv) + tluv + 0x00020002u) >> 2;
+                    YuvToBgr(bottomY[len - 1], (int)(uv0 & 0xff), (int)(uv0 >> 16), bottomDst.Slice((len - 1) * xStep));
+                }
+            }
+        }
+
+#if SUPPORTS_RUNTIME_INTRINSICS
+        // We compute (9*a + 3*b + 3*c + d + 8) / 16 as follows
+        // u = (9*a + 3*b + 3*c + d + 8) / 16
+        //   = (a + (a + 3*b + 3*c + d) / 8 + 1) / 2
+        //   = (a + m + 1) / 2
+        // where m = (a + 3*b + 3*c + d) / 8
+        //         = ((a + b + c + d) / 2 + b + c) / 4
+        //
+        // Let's say  k = (a + b + c + d) / 4.
+        // We can compute k as
+        // k = (s + t + 1) / 2 - ((a^d) | (b^c) | (s^t)) & 1
+        // where s = (a + d + 1) / 2 and t = (b + c + 1) / 2
+        //
+        // Then m can be written as
+        // m = (k + t + 1) / 2 - (((b^c) & (s^t)) | (k^t)) & 1
+        public static void UpSampleSse41(Span<byte> topY, Span<byte> bottomY, Span<byte> topU, Span<byte> topV, Span<byte> curU, Span<byte> curV, Span<byte> topDst, Span<byte> bottomDst, int len, byte[] uvBuffer)
+        {
+            const int xStep = 3;
+            Array.Clear(uvBuffer, 0, uvBuffer.Length);
+            Span<byte> ru = uvBuffer.AsSpan(15);
+            Span<byte> rv = ru.Slice(32);
+
+            // Treat the first pixel in regular way.
+            int uDiag = ((topU[0] + curU[0]) >> 1) + 1;
+            int vDiag = ((topV[0] + curV[0]) >> 1) + 1;
+            int u0t = (topU[0] + uDiag) >> 1;
+            int v0t = (topV[0] + vDiag) >> 1;
+            YuvToBgr(topY[0], u0t, v0t, topDst);
+            if (bottomY != default)
+            {
+                int u0b = (curU[0] + uDiag) >> 1;
+                int v0b = (curV[0] + vDiag) >> 1;
+                YuvToBgr(bottomY[0], u0b, v0b, bottomDst);
+            }
+
+            // For UpSample32Pixels, 17 u/v values must be read-able for each block.
+            int pos;
+            int uvPos;
+            for (pos = 1, uvPos = 0; pos + 32 + 1 <= len; pos += 32, uvPos += 16)
+            {
+                UpSample32Pixels(topU.Slice(uvPos), curU.Slice(uvPos), ru);
+                UpSample32Pixels(topV.Slice(uvPos), curV.Slice(uvPos), rv);
+                ConvertYuvToBgrSse41(topY, bottomY, topDst, bottomDst, ru, rv, pos, xStep);
```

Comment:
I would avoid slicing and passing spans to these private methods here, and iterate instead using reference references or pointers. (Pinning is fine with such complex code IMO if it helps keeping readability.)